### PR TITLE
Removing conversions from Expression.__init__

### PR DESCRIPTION
**Pull request by mmatera**
and more replacements from `leaves` to `elements`.

The nontrivial changes are essentially those proposed in the first patch in #400.

<a href="https://gitpod.io/#https://github.com/Mathics3/mathics-core/pull/401"></a>

### CODE REVIEW

File: mathics/core/pattern.py [hunk 1]
```diff
@@ -355,34 +361,34 @@ def yield_head(head_vars, _):
             # and not OneIdentity &
             # (expression.get_attributes(evaluation.definitions) |
             # expression.get_head().get_attributes(evaluation.definitions)):
-            new_expression = Expression(self.head, expression)
-            for leaf in self.leaves:
-                leaf.match_count = leaf.get_match_count()
-                leaf.candidates = [expression]
-                # leaf.get_match_candidates(
-                #    new_expression.leaves, new_expression, attributes,
+            new_expression = Expression(self.head.expr, expression)
```
Comment (mmatera) [inline mathics/core/pattern.py:364]: Here there is one of the three non trivial changes: `self.head` is a `Pattern` object, but `Expression` needs  a `BaseElement`.

File: mathics/core/expression.py [hunk 2]
```diff
@@ -214,12 +214,27 @@ def __init__(
...
-        # self.elements_properties = elements_properties
+        # self._elements, self.elements_properties = convert_expression_elements(
+        #    elements, lambda x:x
+        # )
+        # assert isinstance(self._elements, tuple)
+        self._elements = elements
+
+        self._build_elements_properties()
+        #
+        # comment mmatera: I found that in certain cases, the elements_properties
```
Comment (mmatera) [inline mathics/core/expression.py:225]: Here is the reason why test_series.py failed.

**Review by TiagoCavalcante — APPROVED** [review]
LGTM

File: mathics/core/pattern.py [hunk 3]
```diff
@@ -131,7 +131,7 @@ def get_elements(self):
         return self.expr.get_elements()
 
     # Compatibily with old code. Deprecated, but remove after a little bit
-    get_leaves = get_elements
+    get_elements = get_elements
```
Comment (rocky) [inline mathics/core/pattern.py:134]: This does nothing.

File: mathics/core/expression.py [hunk 4]
```diff
@@ -199,9 +199,9 @@ def __init__(
     ):
         self.options = None
         self.pattern_sequence = False
-        if isinstance(head, str):
-            # We should fix or convert to to_expression all nonSymbol uses.
-            head = Symbol(head)
+        assert isinstance(head, BaseElement)
```
Comment (rocky) [inline mathics/core/expression.py:202]: Now that this is the case, we should change the signature in line 197 or so to:

`self, head: BaseElement, ...`

File: mathics/core/expression.py [hunk 5]
```diff
@@ -199,9 +199,9 @@ def __init__(
...
+        assert isinstance(elements, tuple)
```
Comment (rocky) [inline mathics/core/expression.py:203]: I have been wavering on this this. When you pass `*x` the result is always a tuple. But I guess it doesn't hurt to check.

Comment (mmatera) [reply mathics/core/expression.py:203]: I put it just to check that in all the platforms and versions, we get the right type. We can comment out these checks now.

File: mathics/core/pattern.py (hunk 3 above)
Comment (mmatera) [reply mathics/core/pattern.py]: There was too many Ctrl+%... I fixed it now.

File: mathics/core/expression.py (hunk 4 above)
Comment (mmatera) [reply mathics/core/expression.py]: Yep. It seems we gain 25% of the time in the doctests avoiding these asserts...

**Comment by rocky** [comment]
LGTM and Thanks!

**Pull request merged**